src/Features/Core/Portable/EditAndContinue/Remote/RemoteEditAndContinueServiceProxy.cs

```diff
@@ -0,0 +1,462 @@
+            /// Remote API.
+            /// </summary>
+            public async Task PrepareModuleForUpdateAsync(Guid mvid, CancellationToken cancellationToken)
+            {
+                try
+                {
+                    await _managedModuleInfoProvider.PrepareModuleForUpdateAsync(mvid, cancellationToken).ConfigureAwait(false);
+                }
+                catch (Exception e) when (FatalError.ReportWithoutCrashUnlessCanceled(e))
+                {
+                    // nop
+                }
+            }
+        }
+
+        public async Task EndEditSessionAsync(IDiagnosticAnalyzerService diagnosticService, CancellationToken cancellationToken)
+        {
+            ImmutableArray<DocumentId> documentsToReanalyze;
+
+            var client = await RemoteHostClient.TryGetClientAsync(_workspace, cancellationToken).ConfigureAwait(false);
+            if (client == null)
+            {
+                GetLocalService().EndEditSession(out documentsToReanalyze);
+            }
+            else
+            {
+                documentsToReanalyze = await client.RunRemoteAsync<ImmutableArray<DocumentId>>(
+                    WellKnownServiceHubService.RemoteEditAndContinueService,
+                    nameof(IRemoteEditAndContinueService.EndEditSessionAsync),
+                    solution: null,
+                    Array.Empty<object>(),
+                    callbackTarget: null,
+                    cancellationToken).ConfigureAwait(false);
+            }
+
+            // clear all reported rude edits:
+            diagnosticService.Reanalyze(_workspace, documentIds: documentsToReanalyze);
+        }
+
+        public async Task EndDebuggingSessionAsync(EditAndContinueDiagnosticUpdateSource diagnosticUpdateSource, IDiagnosticAnalyzerService diagnosticService, CancellationToken cancellationToken)
+        {
+            ImmutableArray<DocumentId> documentsToReanalyze;
+
+            var client = await RemoteHostClient.TryGetClientAsync(_workspace, cancellationToken).ConfigureAwait(false);
+            if (client == null)
+            {
+                GetLocalService().EndDebuggingSession(out documentsToReanalyze);
+            }
+            else
+            {
+                documentsToReanalyze = await client.RunRemoteAsync<ImmutableArray<DocumentId>>(
+                    WellKnownServiceHubService.RemoteEditAndContinueService,
+                    nameof(IRemoteEditAndContinueService.EndDebuggingSessionAsync),
+                    solution: null,
+                    Array.Empty<object>(),
+                    callbackTarget: null,
+                    cancellationToken).ConfigureAwait(false);
+            }
+
+            // clear emit/apply diagnostics reported previously:
+            diagnosticUpdateSource.ClearDiagnostics();
+
+            // clear diagnostics reported during run mode:
+            diagnosticService.Reanalyze(_workspace, documentIds: documentsToReanalyze);
+        }
```

Comment:
minor design point.  it might be nicer to have different services/classes responsible for the 'debuggin lifetiem' code versus the 'session lifetime' code.

Reply:
The proxy is a facade for RemoteEditAndContinueService. Having multiple EnC ISB services would be complicated.